### PR TITLE
PATCH: add back n-dim device-mesh + fix tp trainer saving

### PR DESCRIPTION
**Pull request by S1ro1**
1. Fixes ndim check on device_mesh - This was merged in previously with #38949 but was by mistake reverted by #39501, we need this for upcoming accelerate/axolotl release. 
2. Makes sure we save properly on distributed rank if tp allowed in trainer

cc @ArthurZucker @SunMarc

### CODE REVIEW

File: src/transformers/modeling_utils.py [hunk 1]
```diff
@@ -4617,10 +4617,14 @@ def from_pretrained(
...
             else:
-                # TODO: make device_mesh support multiple dimensions
-                if device_mesh.ndim > 1:
-                    raise ValueError("device_mesh must be 1 dimensional and will be used for TP")
-                device_map = torch.device(device_mesh.device_type, int(os.environ["LOCAL_RANK"]))
+                if "tp" not in device_mesh.mesh_dim_names:
+                    raise ValueError(
+                        "When using `tp_plan`, the `device_mesh` must contain a 'tp' dimension. "
+                        "Please provide a valid `device_mesh`."
+                    )
```
Comment (ArthurZucker) [inline src/transformers/modeling_utils.py]: I don't think we should enforce `'tp'` in the device mesh! for inference we never use that!~

File: src/transformers/modeling_utils.py [hunk 2]
```diff
@@ -4617,10 +4617,14 @@ def from_pretrained(
...
+                device_mesh = device_mesh["tp"]
+                tp_size = device_mesh["tp"].size()
+                device_map = torch.device(f"{device_mesh.device_type}:{int(os.environ['LOCAL_RANK'])}")
```
Comment (ArthurZucker) [inline src/transformers/modeling_utils.py]: only do this if `tp` exists in it!

File: src/transformers/trainer.py [hunk 3]
```diff
@@ -3953,6 +3953,13 @@ def save_model(self, output_dir: Optional[str] = None, _internal_call: bool = Fa
             if IS_SAGEMAKER_MP_POST_1_10:
                 # 'user_content.pt' indicates model state_dict saved with smp >= 1.10
                 Path(os.path.join(output_dir, "user_content.pt")).touch()
+        # We are in N-D parallelism if we have parallelism_config set, so we check accelerate if we're on a to_save rank
+        elif (getattr(self.accelerator, "parallelism_config")) is not None:
```
Comment (ArthurZucker) [inline src/transformers/trainer.py]: okay!

File: src/transformers/modeling_utils.py [hunk 4]
```diff
@@ -4617,10 +4617,15 @@ def from_pretrained(
...
-                # TODO: make device_mesh support multiple dimensions
                 if device_mesh.ndim > 1:
-                    raise ValueError("device_mesh must be 1 dimensional and will be used for TP")
-                device_map = torch.device(device_mesh.device_type, int(os.environ["LOCAL_RANK"]))
+                    if "tp" not in device_mesh.mesh_dim_names:
+                        raise ValueError(
+                            "When using `tp_plan`, the `device_mesh` must contain a 'tp' dimension. "
+                            "Please provide a valid `device_mesh`."
+                        )
+                    device_mesh = device_mesh["tp"]
```
Comment (ArthurZucker) [inline src/transformers/modeling_utils.py]: can it be ndim > 1 but not mesh dim names?

**Review by ArthurZucker — APPROVED** [review]
Thanks!

File: src/transformers/modeling_utils.py (hunk 4 above)
Comment (S1ro1) [reply src/transformers/modeling_utils.py]: Nope it can't IMO, how do you think we should take the correct submesh then?

**Comment by S1ro1** [comment]
Fails unrelated, merging

File: src/transformers/modeling_utils.py (hunk 4 above)
Comment (ArthurZucker) [reply src/transformers/modeling_utils.py]: No no I just want to be sure as the basic initialization we do is without providing a mesh name!

Comment (S1ro1) [reply src/transformers/modeling_utils.py]: Oh, that works, we check for "tp" only if ndim > 1, therefore the basic initialization still works. We check for "tp" only if mesh.ndim > 1 AND user-provided mesh. If this passes we select the correct submesh ("tp") and use that as 1D mesh afterwards, as if it was created by us in `initialize_tensor_parallelism`
